### ISSUE TITLE
Bug: Hydration mismatch error due to plugins generating script tag on top

### ISSUE THREAD

**Issue by yongdamsh** [facebook/react#24430]
React version: 18.0.0, 18.1.0-next-fc47cb1b6-20220404 (latest version in codesandbox)

## Steps To Reproduce

1. Install a plugin that creates a script tag at the top(ex: [Apollo Client Devtools](https://chrome.google.com/webstore/detail/apollo-client-devtools/jdkknkkbebbapilgoeccciglkfbmbnfm))
2. Go to the [demo](https://codesandbox.io/s/kind-sammet-j56ro?file=/src/App.js) in the [new SSR suspense guide](https://github.com/reactwg/react-18/discussions/37)
3. Open preview in a new window
4. UI mismatch error occurs at hydration time

  <img width="600" alt="스크린샷 2022-04-24 오전 11 02 34" src="https://user-images.githubusercontent.com/4126644/164952677-06618e72-3343-4b92-9eaa-b45bce11c3ab.png">

Link to code example: https://codesandbox.io/s/kind-sammet-j56ro?file=/src/App.js

## The current behavior

If a script tag is inserted before the head tag due to the user's browser environment such as a plugin, it is judged as a hydration mismatch and the screen is broken.

https://user-images.githubusercontent.com/4126644/164953071-14546c74-d9ab-4a6f-8f99-6712f29c6dd6.mov

## The expected behavior

This problem may be a part that each third party needs to solve, but I'm wondering if it's possible to handle an exception in the hydration matching logic of React.

**Comment by gaearon** [comment]
Might be related to (or same as) https://github.com/facebook/react/issues/22833, but let's keep both open for now

**Comment by marcusthelin** [comment]
Is the `hydrateRoot` function expecting the whole html document to be exactly what gets rendered by `renderToPipeableStream`? The best would be to just try to hydrate the React root element where the app is rendered? 

It does seem that to use `renderToPipeableStream` i need to render the whole HTML document with a React component, but this is not ideal when e.g. using Vite with SSR in development, since it needs transform the html to inject custom scripts.

**Comment by hrgui** [comment]
From my understanding, if anything else other than document was passed in into hydrateRoot, it doesn't seem to crash when I have chrome extensions that modify the DOM installed (e.g. Dark Reader / Apollo DevTools). 

In the above example, I changed the following:

Html.js
```diff
export default function Html({ assets, children, title }) {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="shortcut icon" href="favicon.ico" />
        <link rel="stylesheet" href={assets["main.css"]} />
        <title>{title}</title>
      </head>
      <body>
        <noscript
          dangerouslySetInnerHTML={{
            __html: `<b>Enable JavaScript to run this app.</b>`
          }}
        />
+        <div id="root">{children}</div>
-       {children}
        <script
          dangerouslySetInnerHTML={{
            __html: `assetManifest = ${JSON.stringify(assets)};`
          }}
        />
      </body>
    </html>
  );
}
```

App.js

```diff
export default function App({ assets }) {
  return (
    <Html assets={assets} title="Hello">
+         <AppContent />
-         <Suspense fallback={<Spinner />}>
-            <ErrorBoundary FallbackComponent={Error}>
-                <Content />
-            </ErrorBoundary>
=         </Suspense>
    </Html>
  );
}

+ export function AppContent() {
+  return (
+    <Suspense fallback={<Spinner />}>
+      <ErrorBoundary FallbackComponent={Error}>
+        <Content />
+      </ErrorBoundary>
+    </Suspense>
+  );
+ }
```


index.js:
```diff
- hydrateRoot(document <AppContent />);
+ hydrateRoot(document.getElementById("root"), <AppContent />);
```

---
I don't know if this crashes with Cypress though.

**Comment by gaearon** [comment]
I don’t think the “stricter” behavior here is intentional. I’ll be checking with the team but my current understanding is that this is a bug.

**Comment by hrgui** [comment]
Something I've noticed is that when React does encounter a hydration mismatch, it attempts to fallback to client side rendering.

Which does show up in the example codesandbox (the one where we are hydrating the document):
```
Warning: An error occurred during hydration. The server HTML was replaced with client content in <#document>.
```

```
Uncaught Error: There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.
```

However, it results in an application crash because of the next error:

```
 Failed to execute 'appendChild' on 'Node': Only one element on document allowed.
```

which is from the call stack:
appendChildToContainer <- insertOrAppendPlacementNodeIntoContainer (3) <- commitPlacement <- commitMutationEffectsOnFiber <- commitMutationEffects_complete <- commitMutationEffects_begin <- commitMutationEffects <- commitRootImpl

Then later another error is thrown:

```
Uncaught DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
```

which is from the call stack:
removeChildFromContainer <- unmountHostComponents <- commitDeletion <- commitMutationEffects_begin <- commitMutationEffects <- commitRootImpl <- commitRoot <- performSyncWorkOnRoot <- flushSyncCallbacks <- commitRootImpl

---

What I am wondering about is: In the case of falling back to client side rendering, why does React do `appendChild` instead of `replaceChild`? If it was `replaceChild`, there wouldn't be an app crash in this case.

**Comment by gaearon** [comment]
Yes, that issue is https://github.com/facebook/react/issues/22833. I believe the fix we wanted to do is changing which top-level element we modify. (Maybe body instead of html?) It would be nice not to have to introduce a separate “host config” method (which we’d have to do if we added a call to “replace”). So ideally the fix should use the methods we already use.

**Comment by rdadoune** [comment]
Not an ideal solution but in my use case, I'm only concerned with generating/modifying the `head` during SSR, and the following hack allows me to work around errors that occur as the result of modifications to the `head` outside of React by Cypress, various chrome plugins, etc. 

```typescript
const Head: React.FC = () => {

  if (globalThis.document?.head) {
    return (
      <head dangerouslySetInnerHTML={{ __html: document.head.innerHTML }} />
    );
  }

  return (
    <head>
      {/* ... Do important things on the server */}
    </head>
  );
};
```

This is especially useful for me because even with current fixes added to `react-dom@next` that allow the client to "recover", doing so wipes out all styles generated by `@emotion` that have been collected into the `head`.

**Comment by Mordred** [comment]
I tried React 18.2.0, but it still breaks the page when Apollo Client DevTools extension is used. So my _ugly_ solution for this problem is fix the DOM before `hydrateRoot`:

```js
document.querySelectorAll('html > script').forEach((s) => {
  s.parentNode.removeChild(s);
});

const root = hydrateRoot(
  document,
  <React.StrictMode>
    <AppWithFullDocument />
  </React.StrictMode>,
);
```

Note: You can replace the selector more strict query .. e.g. `'html > script[src^=chrome-extension]'`

(remix-run/remix#2947)

**Comment by dbashford** [comment]
Getting the issue with Cypress tests, as suggested in Remix Discord, my workaround for those coming here:

```jsx
if (process.env.NODE_ENV === 'test') {
  require('react-dom').hydrate(<RemixBrowser />, document);
} else {
  hydrateRoot(document, <RemixBrowser />);
}
```

🙏 for a fix soon, thx all!

**Comment by gaearon** [comment]
@dbashford Which in particular issue are you hitting? We’ve released a short-term fix for the most urgent part of the problem in 18.2 (document fallback didn’t work at all). Now it should work by doing a full client render but this will lose the third party things. There is a possible fix to do something smarter but it’s a bigger project and a bit further on the timeline. So it would help to know what exactly broke for you. Is there a repro? Particular scenario? Thanks.

**Comment by dbashford** [comment]
Mine is the Cypress case. Everything works fine in dev and in prod, but Cypress when it kicks up fails (418), and only after making the `hydrateRoot` switch. I haven't dug into it too far to really understand the problem, but I've read in this thread and in the Remix discord that the problem may be that Cypress monkey's with the head causing the hydration issue. I admit that SSR/hydration are still voodoo/witchcraft to me, so I'm struggling a bit to work my way through it. Baby steps.

**Comment by CanRau** [comment]
Hey @Mordred thanks a lot for this quick hack, I modified it slightly to also get rid of `input`s within `html` as the Yoroi extensions injects a hidden input directly into the `<html/>` (not `<body/>`)

document.querySelectorAll("html > script, html > input").forEach((s) => {
  s.parentNode?.removeChild(s);
});

**Comment by gaearon** [comment]
@dbashford Can you share a minimal project? It always helps to have concrete examples to check fixes against. This goes for everyone else too.

**Comment by dbashford** [comment]
@gaearon Can call off the dogs on that latest post, was something I introduced myself while trying to debug a real hydration issue with dates.

The Cypress issue is still a problem, that's all local and consistent. I'll work to get a repro up over the next few weeks when I get a second.

**Comment by adbutterfield** [comment]
I just updated my repo with some instructions to more easily reproduce the issue, and test that it works when not using styled components.

See here: https://github.com/adbutterfield/fast-refresh-express/tree/react-18

**Comment by camjackson** [comment]
Hi @gaearon I have a repro repo for the hydration errors with cypress. It's not exactly "minimal", it's a freshly-created [Remix grunge stack](https://github.com/remix-run/grunge-stack) app, which has quite a bit of stuff in it. Here it is: https://github.com/camjackson/remix-cypress-error.

Here are the steps to repro the error locally:
1. `git clone git@github.com:camjackson/remix-cypress-error.git`
2. `cd remix-cypress-error`
3. `yarn`
4. `yarn test:e2e:run`

The test will fail with a hydration error. To show that it's related to react 18, you can open up `app/entry.client.tsx` and comment/un-comment a few lines to switch it from `hydrateRoot` back to the old `hydrate`. Then run the e2es again and it will pass.

To debug it further, instead of `yarn test:e2e:run` you can do `yarn test:e2e:dev` to fire up cypress in interactive mode and see the test fail in a real browser. 

Oh and one final note, if you just start the app normally with `yarn dev` you should see that there are no errors. It only happens with cypress, presumably because cypress injects something extra into the document.

<details>
<summary>For completeness, here's how I created this repo from scratch:</summary>

1. `yarn create remix --template remix-run/grunge-stack`
6. Go through the prompts, give it a name and `cd` into the newly created project
7. `yarn add react react-dom`
8. `yarn add -D @types/react @types/react-dom`
9. Open up `app/entry.client.tsx` and migrate from `react-dom`'s `hydrate` to `react-dom/client`'s `hydrateRoot`
</details>

**Comment by pegak** [comment]
I have replicated same behaviour - it can be replicated anywhere with `hydrateRoot` with any Chrome extension that injects the script. It then switches to client-side rendering.

`Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.`

**Comment by 0Lucifer0** [comment]
it's also preventing loom (extension) to work as the extension add a to the dom. Seems like it just break any extension that update the dom and having a workaround every potential extension is going to be a nightmare

fix for loom

document.querySelectorAll('loom-container').forEach((s) => {
    s.parentNode.removeChild(s);
});

additionally this also cause warning if the extension is adding attributes like this one with colorzilla 
![image](https://user-images.githubusercontent.com/15112753/175883147-39aa7b94-154b-4888-afa8-a2e475c4753a.png)

**Comment by dbashford** [comment]
Also causing issue with LastPass, can confirm Loom as well, causing all issues across multiple plugins for us.

**Comment by SokratisVidros** [comment]
Apart from extensions, this issue is affecting JS libraries that hot load scripts or styles dynamically on bootstrapping.  In [Clerk](https://clerk.dev/) we use this technique to ensure that users of Clerk-powered applications always have the latest fixes in their authentication components.

A workaround is to delay the hot loading of the components code on the client side. Unfortunately, this is not acceptable for authentication as it will increase the necessary load time before the user is able to interact with the page.

**Comment by dejanr** [comment]
Still an issue for react apps that render full document. If anyone is using browser extension that injects scripts during the page load, there is an hydration error occuring. Is it safe to use hydrate instead of hydrateRoot, since someone mentioned that it is working with react-dom hydrate as intended.

**Comment by kamranayub** [comment]
Happening with Grammarly and Vimeo Recorder too 😞 If I disable LastPass, and the other two, it's fine.

**Comment by Ryongyon** [comment]
This problem makes me unable to use Remix framework normally. Although we know that disabling Chrome extensions can solve the problem, we cannot ask our users to disable Chrome extensions. However, many Chrome extensions will inject content into Html documents now, so I wonder when the authorities will pay attention to this problem?

**Comment by denchiklut** [comment]
@gaearon It looks like the problem appears only when u try to stream html into `document` 
because of additional scripts generated on server for partial hydration. So when I tried to stream the app in the div#root it really worked fine and solve all hydration issues.

This this the [working code example (branch: feat/stream)](https://github.com/denchiklut/ssr-boilerplate/blob/feat/stream/src/server/middleware/render/index.tsx#L19)


And another think I'd like to ask if there will be a way to `Transform` stream in order to have more control over generated html?

ideally it would be nice to have something like this:
```jsx
const { pipe } = renderToPipeableStream(jsx, {
  bootstrapScripts: chunkExtractor.getMainAssets('script').map(i => i.url),
    onShellReady() {
        res.setHeader('Content-type', 'text/html')
        res.write(getHtml(chunkExtractor))
        pipe(new Transform({
        // ... do some manipulation here,
        })
        pipe(res)
    }
  })
```

**Comment by 0Lucifer0** [comment]
> I was able to resolve my issue by invoking [`clearBrowserExtensionInjectionsBeforeHydration()`](https://gist.github.com/OnurGvnc/31f03f0d5237b78224aa083493fda645#file-clearbrowserextensioninjectionsbeforehydration-ts-L5-L40) prior to executing `hydrateRoot()`. https://gist.github.com/OnurGvnc/31f03f0d5237b78224aa083493fda645

Yeah it's not really a good fix as it just remove all the potential dom changes... Still useful as a temp fix until better solution... not sure why this bug is still tagged as unconfirmed after 9 month when it's literally making react18 not production ready

**Comment by Mihai-github** [comment]
Hi, still having issues with this ... I saw/read a lot of issues and did not see a universal solution (if there is one) ... maybe someone can help me with an answer?
Also on a side note... not sure it is related but I have CSS flickers and not sure if is related to this or not.

Thanks.

**Comment by erwan-joly** [comment]
> Hi, still having issues with this ... I saw/read a lot of issues and did not see a universal solution (if there is one) ... maybe someone can help me with an answer? Also on a side note... not sure it is related but I have CSS flickers and not sure if is related to this or not.
> 
> Thanks.

Same for the css flickers not sure why. I don’t think it’s related as I didn’t use to have it it appeared recently

**Comment by Mihai-github** [comment]
@erwan-joly It's related to the hydration because I switched for testing this to react v17 and the flickers disappeared.

**Comment by Ryongyon** [comment]
Expect to be more of a Chrome problem than a React problem

In my research and observation, I have found that Chrome extensions that inject content into a page occasionally trigger this error, such as "React Developer Tools," which occasionally causes hydration errors. Later I read that Chrome extensions do not always inject content into a page at a fixed time, which means sometimes before the page is loaded, sometimes after the page is loaded. If you inject after the page is loaded, you won't get an error. If you inject before the page is loaded, you will naturally fail to hydrate.

Once people bring up the subject, Remix officials will just say it's a React problem and do nothing about it. Why is Remix the only SSR framework with this problem? Or does Remix just not care about the developer's feelings? Remember that most browser extensions now inject scripted content into the page, and if this isn't fixed, it means Remix won't work in production.

While we know that disabling Chrome extensions can fix this, we can't ask our users to disable Chrome extensions

I believe the reason the Remix website doesn't have this issue is because they are using the React 17 version, if they are using the latest version they will have this issue

**Comment by martdavidson** [comment]
> Why is Remix the only SSR framework with this problem?

It's not, I've just spent an afternoon debugging why our production build was doing this with next.js - because I have lastpass installed, and it was injecting it's little icon into a username field.

Would love to see this at least get the `Status: Unconfirmed` removed, and get at least an idea of what is being thought about behind the scenes, it really seems to defeat a lot of the benefits of SSR with hydration if not I'm not misunderstanding.

**Comment by Aprillion** [comment]
> Expect to be more of a Chrome problem than a React problem
> 
> In my research and observation, I have found that Chrome extensions that inject content into a page occasionally trigger this error...

@Ryongyon  unfortunately, there is nothing that Chrome/Firefox extension makers can do about it, and all of those extensions coexisted with React for many years, so it's React 18 that made this problem materialize and thus it's up to React to fix it, we cannot expect thousands of extension makers to figure out how to deal with sudden influx of rogue React websites

e.g. https://github.com/darkreader/darkreader/issues/8842#issuecomment-1120342766

**Comment by Aprillion** [comment]
@gnoff I see you are assigned to this issue since June 2022, but it still has `Status: Unconfirmed` even after we provided reproduction steps and many people confirmed it's a problem with many SSR frameworks and many browser extensions ... How can the community help more, please?

**Comment by Xiphe** [comment]
Also running into this issue. In my case, the worst effect is due to react rebuilding the document, the injected `<style />`-tag from styled-components is being removed and the lib does not recover from this state (ref https://github.com/styled-components/styled-components/issues/3924)

Is there a way to tell react not to hydrate a nested part of the app at all? (`suppressHydrationWarning` will still overwrite the original contents) like `<div><span>React manages this</span><span endOfReact={true}>React won't care about this</span></div>`?

**Comment by cayblood** [comment]
> I experience the same issue with my Remix app and have to use `hydrate` instead of `hydrateRoot` to make it work properly.

Can someone explain the tradeoff of using this workaround?

**Comment by 0Lucifer0** [comment]
> > I experience the same issue with my Remix app and have to use `hydrate` instead of `hydrateRoot` to make it work properly.
> 
> Can someone explain the tradeoff of using this workaround?

hydrate is deprecated as replaced by hydrateRoot. 
My understanding is that the main difference is around streaming SSR. I believe streaming SSR won't be possible with hydrate. It's pretty much like if you were running react 17.

**Comment by cayblood** [comment]
Remix folks, look no further than this repo for a React 18-compatible workaround:
https://github.com/kiliman/remix-hydration-fix

**Comment by gaearon** [comment]
>I see you are assigned to this issue since June 2022, but it still has Status: Unconfirmed

Just wanted to jump in here real quick — @gnoff has been focusing on different related things in this area for a while (several months). We just don't use issue labels a lot, so we don't always update them. From what I understand, we already fixed some issues in main (available as `@next` canary prereleases), and for others there is a bigger ongoing project. But I don't have exact details yet so I'll ask around.

**Comment by gaearon** [comment]
Regarding the fixes that already landed in main, it would help if you could confirm which of these issues still occur with `react@next` and `react-dom@next` prereleases.

**Comment by adbutterfield** [comment]
Problem remains with react@next and react-dom@next on my repo: https://github.com/adbutterfield/fast-refresh-express/tree/react-18

**Comment by hrgui** [comment]
https://github.com/hrgui/react-next-hydrateRoot-problem. I've updated the package deps to use `react@next` and `react-dom@next`

When Dark Reader runs on my example on http://localhost:4000, the server renders, then React checks if the client matches. Since Dark Reader has changes, I still see it re-render since 

```
Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.
```

```
An error occurred during hydration. The server HTML was replaced with client content in <#document>.
```

**Comment by gnoff** [comment]
Hey folks, here's a brief update on what I've been working on and what you can expect to be addressed in the next stable release of React.

This issue has attracted a wide array of failure modes related to hydration. Originally it was centered around trying to SSR the entire document and then call `hydrateRoot(document,...)` on the client. Since then the issues have expanded to include problems with any browser extension modifying the DOM before React has a chance to hydrate it. What I'd like to do is explain the current constraints of hydration, what should be fixed in the next stable, and what will continue to be problematic interop with extensions that we may or may not try to solve.

### How hydration works today
When React hydrates your application it looks in the DOM for the next element that should have a partner host element in your app. If it sees a `<div>` it will expect the next html your apps renders to be a `<div>` as well. As you can imagine if an extension injects a `<script>` tag in the middle of your SSR'd HTML then this hydration algorithm is going to stumble upon it and expect you to render that `<script>` and when you don't (because it was injected by a browser extension) hydration will fail and some of your React app will fall back to client rendering where the DOM will be discarded and a new one will be constructed from the render).

React has always found hydrating `<html>`, `<head>`, and `<body>` tags challenging because these are the most commonly targeted DOM nodes for manipulation by 3rd party scripts and extensions and that is why for a long while React did not recommend rendering from the `Document` but instead use some `<div id="my-root" />` embedded in the body. One affordance made though was that for these tags specifically React would ignore any extra Nodes it found when hydrating so even if a 3rd party script or extension did inject some extra DOM Nodes, hydration would continue without trying to hydrate them.

Another interop challenge is that when React falls back to client rendering, and it is hydrating the `Document` then it will remove the `<html>` Node and all children while it prepares the client rendered nodes to re-insert. This will remove 3rd party styles from the DOM and while most 3rd party stuff is resilient to this (it's a hostile environment, trying to patch someone elses HTML) they aren't all and this can break extensions in ways we don't like.

This is the state of React and hydration today and I believe the cause of many but not all of the issues reported in this issue so far. To summarize, React expects the HTML you render to match what is in the DOM in the same order and for `<head>` and `<body>` it is a little more tolerant of extra nodes appearing AFTER your React Nodes which it ignores. It might also remove and recreate the entire DOM which may not play nice with 3rd party scripts and extensions.

### What is New
In the next stable a number of new features are going to be included which should alleviate some of the pain points above.

1. React will treat `<html>`, `<head>`, and `<body>` as special types of host elements and will never remove them from the DOM. Even if React has to render the whold document from scratch, it will only remove the contents of these elements, preserving style/stylesheets (scripts don't need to be preserved because once they are in the DOM they execute regardless of whether they are removed or not)

2. React will start to treat certain tags as hoistable. Things like `<title>`, `<meta>`, and `<link>` will under most circumstances be hoisted to the `<head> independent of their rendered position in the Document. Along with these new hoisting semantics are new hydration semantics. Position in the Document won't govern a hydration match here so if the <head> has extra nodes injected in random places there should be limited impact.

3. React will start to treat certain kinds of browser resources like stylesheets (with an opt-in) and async scripts (with an opt-out) as hositable and dedupable. I'll share more about this in a proper post somewhere else but similar to hoistable tags if you opt into resource semantics hydration will not use the normal linear flow

In summary, the vast majority of tags that get injected into `<head>` and `<body>` should avoid the normal linear flow of hydration and make collisions with your app's hydratable elements much harder to encounter.

### What is not changing yet
React still expects normal (non-hoisted) tags to appear in the order rendered on the server without any intervening HTML Nodes added by 3rd party scripts and browser extensions. We also have not done anything to address hydration mismatches due to extensions modifying nodes deep in your body. For instance, in one of the referenced cases, a password manager adds an icon to an input field which breaks hydration. This is probably more of a prevalent problem in React 18 than in prior versions because hydration falls back to client rendering rather than patching up the DOM as it hydrates. We will consider if there are options for improving this situation in the future but this is not addressed by my work.

**Comment by gnoff** [comment]
>When Dark Reader runs on my example on http://localhost:4000/, the server renders, then React checks if the client matches. Since Dark Reader has changes, I still see it re-render the entire document with the following error messages:

@hrgui thanks for checking that out. I took a look at this extension and it does inject sync scripts and style tags at the beginning of the <head>. This is surprising in some regard (there really isn't a reason to insert a script anywhere in particular) though the <style> perhaps requires some kind of precedence shadowing. What I can say is that given the contraints I outlined above it is expected that this extension interferes with hydration as it currently is implemented. I will consider what we could do to be more tolerant here

**Comment by gaearon** [comment]
>Regarding the fixes that already landed in main, it would help if you could confirm which of these issues still occur with react@next and react-dom@next prereleases.

As a related note — it would also be good to check for each case that the causes of the problem are genuinely new in 18, and what _exactly_ the root cause is in each case. E.g. is it related to streaming (initial hydration no longer means all HTML has arrived), to non-blocking hydration (`startTransition` and `Suspense` let us do it in chunks instead of blocking for the entire page), to hydrating the whole document, or to React's hardened behavior on mismatches when something mismatches (it now re-renders up to the closest `Suspense` boundary on the client instead of leaving the document in an inconsistent state).

I know it's not very easy to determine but I think it's really important to categorize each case so that we know where the bulk of the issues is. In some cases it could be easy userland fixes — e.g. deferring some script from executing until a certain point. Or telling React not to touch content controlled by something else with `suppressHydrationWarning` or even `dangerouslySetInnerHTML`. Or relaxing some heuristic in React so that it's less strict in cases where inconsistencies don't risk security bugs.

---

To give you an example, here's an example with a hydration that started appearing in 18: https://github.com/dabit3/nextjs-appdir-embed-testing/. It renders Twitter embed code (which has an HTML snippet and a script tag that would replace it with Twitter's client-rendered embed). The issue *started* happening because Next.js (thanks to React 18) can hydrate without blocking the page. It's better! But as a result, the Twitter embed code ran "too early" and modified the HTML before React hydrated. This caused the error.

My first instinct was to delay running the third-party script (see https://github.com/dabit3/nextjs-appdir-embed-testing/pull/1/commits/1ca23b6cb5190f861c0541a42907718b2dd2843c, Next's `<Script>` component does that by default). But actually, that was not fundamentally the right solution. Again, the problem is that the script *replaced a piece of DOM owned by React*. Even if this happened after hydration, React would *still* get confused if it tried to later update this part of the page during a state update. It's just that now this shows up as a hydration error (recovering with a clean client re-render) rather than a later error.

The actual correct solution *in this case* was to tell React that this part of the page is not owned by React — and so it shouldn't worry about it (see https://github.com/dabit3/nextjs-appdir-embed-testing/pull/1/commits/ab48f48de16294b1aa48777fc1f8b2bdcc6c2bab). This would prevent future client-side errors too because React would skip over that HTML. 

---

This is to illustrate that grouping all these errors as "React 18 errors" makes it easy to miss the root of the problem. For each root problem, we'd like to either provide a fix, a workaround, or at least an explanation for *why exactly* this breaks, so that you can make an informed choice about some tradeoff. (E.g. if it's streaming in particular that broke something, you could disable that, and so on.) I can't promise we'll fix all of these cases — we likely won't — but the more information we have about the actual root causes, the better recommendations we will be able to give. And hopefully fix a few things in React too.

**Comment by gaearon** [comment]
>We also have not done anything to address hydration mismatches due to extensions modifying nodes deep in your body. For instance, in one of the referenced cases, a password manager adds an icon to an input field which breaks hydration. This is probably more of a prevalent problem in React 18 than in prior versions because hydration falls back to client rendering rather than patching up the DOM as it hydrates.

Btw, I want to emphasize the other side of this as well. React 18 *is* more conservative about hydration errors in the sense that it re-renders the closest Suspense boundary with a clean client render, ignoring server HTML. While this is not very efficient, it is safer than what React did before (trying to "guess" which nodes to delete and update as it patches up the tree).

As a concrete example, suppose you're rendering a list of chat threads in a messenger app. They have avatars and names. Suppose you have a bug, and the user list on the server is in a slightly different order than when hydrated on the client.

React 17 would patch up each user's name to match the order from the client. However, it wouldn't patch up their attributes (it's too slow to iterate over each every time) — so their avatars would not match their names. Their profile links would also not match. If you're not careful, this could lead you to messaging a wrong user by mistake. And in any case, as a user, you wouldn't know *which* information is right. Do you trust the name, the avatar, or the link?

React 18 would not attempt to do this patch-up. It would blow away the server HTML up to the closest `<Suspense>` boundary (which would probably be somewhere around the list of users) and render that list on the client from scratch. That's a security/consistency feature. Now all links and avatars and names match up. Of course, it's not perfect — this is slower (in the bad case). It also doesn't catch all possible inconsistencies (like if only attributes were different). But I wanted to explain why this is a feature, not a bug. With granular `<Suspense>` boundaries and a way to opt-out for specific cases like timestamps via `suppressHydrationWarning`, this behavior offers a pretty good balance.

**Comment by gaearon** [comment]
I also wanted to add that we don't recommend to fix this by hydrating independent divs. At least not in long term. Hydrating the full document *is* the right strategy, and the work @gnoff described in https://github.com/facebook/react/issues/24430#issuecomment-1440427646 should help with that. But again, we need to look at each problem on a case-by-case basis. Different problems need different solutions.

**Comment by scerelli** [comment]
I also have the same issue using Next 13 and React 18. 

This is my scenario:
- React Server Component page
- Inside the page i have a client side rendered form with some input fields
- In my browser I have Dashlane extension (password manager that adds the auto-fill functionality for forms),
- Page breaks with: `Error: Hydration failed because the initial UI does not match what was rendered on the server.`  

The error doesn't happen in Incognito or if i disable Dashlane extension. It also happen with other extensions.

**Comment by gnoff** [comment]
@scerelli I tried to see if this extenion's source code is available but wasn't able to find it on github. Do you know if it is reviewable?

Also, when the hydration fails is the entire app broken or does the app recover and there is just an error in the console?

Can you link a repro?

**Comment by scerelli** [comment]
@gnoff here you go: https://github.com/scerelli/hydration-issue-reproduction
![Screenshot 2023-03-08 at 21 04 37](https://user-images.githubusercontent.com/4821190/223835486-9ea0ca0c-9c91-4066-9cc1-d7e1a36fead6.png)

Simply install the Dashlane extension and create a free account. However, this issue may also occur with any other browser extension that modifies the DOM.

I have observed that in my situation, the problem occurs when I am using Yup with React Hook Form. However, I am uncertain about the cause of the problem since the component that is displaying the issue is also a client component. Interestingly, on a Next 12 setup, the same code does not trigger the error.

**Comment by mdodge-ecgrow** [comment]
I'll be following this issue closely.  I was attempting to create and app using remix.run and couldn't even get an initial basic build working because of this issue and I run a lot of browser extensions.

Just me thinking out loud here, but it's too bad the browsers can't enforce some sort of standard for extensions for how they get to manipulate the head and/or body content.  Seems to me if there was that set standard, it would be a lot easier to React to look for and account for those changes.

**Comment by gnoff** [comment]
@scerelli thanks for the repro. I was also hopign to see the dashlane source but I can tell they are injecting dom nodes and adding attributes deep within the application.

React was not updated to be more tolerant of deep mismatches and instead focussed on handling cases in the head and body more effectively. For a lot of 3rd party stuff this is sufficient because that's all they do. But whenever an extension modifies the HTML deeply in the document React can't tell whether that was because you rendered different stuff on server/client or whether it was intentional.

The reason this was not a probem in React 17 is that version just does an insert and patches up the DOM rather than falling back to client rendering.

The fact that you get a bunch of big red errors is offputting and we should figure out a way to surface meaningful info for devs in dev mode to figure out if their app is not correctly hydrating in the baseline case but not cause undo worry when there are incidental hydration errors which are always going to happen especially for users that use 3rd parties that modify the DOM.

The good news is that your repro demonstrates that React can now effectively recover from a hydration mismatch even when the entire document is rendered by it. Before this would have wrecked the whole app because the html element would be unmounted in an ineffective way and it would fail to remount correctly.

I'll continue to gather info from these reports and may someday pursue a more ambitous look at coexisting with extensions deep in the DOM but I don't have anything planned for that at the moment. Thanks for taking the time to report this

**Comment by scerelli** [comment]
@gnoff Thanks to you for taking the time to explain further what's going on. Great work!

**Comment by ErwinAI** [comment]
Hello @gnoff @gaearon  ,

**Tldr;** I am a developer of a dev tool chrome extension (Tailscan) and thus far, three of my users have reported with Remix / React 18. I've been following this issue for a while now and want to chip in, just to give an extra insight in the issue's occurrance. I don't use React / Remix myself, but if it's useful, I can give you access to my private repository containing the chrome extension's code and a packaged version, _just in case you want or need to dive in deeper._

In any case, I very much appreciate the extensive information both of you have provided, thank you!


Each of the three users have ran into the same problems as others have described. It's either hydration errors or the following:

One user reported the following error:

> `Warning: Did not expect server HTML to contain a <tailscan-container> in <html>.`

Another user wrote this:

> Changing the browser extension settings to only load the extension when clicking on the extension icon, instead of automatically on page load, fixes the problem on first page load. If I refresh the page a lot of times I can be lucky that the extension loads after the page is loaded (the code for the service worker pops up in the source tab in the dev tools a little delayed when this happens). My react version is set to "^18.2.0"

Furthermore, that same user reports:

> I can consistently make the extension work by doing the following:
> 1. Set site access to "On click" (_referring to the aforementioned setting. In Chrome, right click the extension icon > "this can read and change site data" > "when you click the extension"_)
> 2. Access the site in a newly open tab
> If I refresh the tab (not opening a new tab) it gets the hydration error where Tailscan won't load.
> 3. If I open a new tab again Tailscan will work without any problems even though I navigate around, refresh the page etc. It is only after Tailscan has been initialized/loaded the first time, it breaks on subsequent refreshes.

The chrome extension does a couple of things within the DOM:
1. There is a <tailscan-container> div element inserted containing a shadow dom underneath the <body> tag
2. A specific class is added to an element that is being inspected (_sidenote: removed in next version_)
3. There are two attributes added to an element that is being inspected (`data-tailscan-id` and `toggled-off-classes`)
4. There are <style> elements inserted at the bottom of the <head> element
5. Add and remove classes on an element that is being inspected

The way this chrome extension works is very similar to other extensions such as password managers inserting things in <input> elements, other devtool extensions, etc.

I think the explanation of the users match up with the explanation @gaearon gave. Furthermore, the changes in the next stable that @gnoff mentions, will help solve some the issues I have. Two unknown for me are still:
1. If custom (data) attributes added to elements in the <body> element will be ignored or respected, or if that will still cause issues in the next stable.
2. If changing classes on an element in the <body> tag will still cause issues in the next stable.

**Comment by 19Qingfeng** [comment]
I want to know the result of this problem. I can't restrict the user's behavior to cause this problem.

**Comment by roggc** [comment]
> Hey @Mordred thanks a lot for this quick hack, I modified it slightly to also get rid of `input`s within `html` as the Yoroi extensions injects a hidden input directly into the `<html/>` (not `<body/>`)
> 
> ```ts
> document.querySelectorAll("html > script, html > input").forEach((s) => {
>   s.parentNode?.removeChild(s);
> });
> ```
> 
> > Using this now in my [Remix Deno Stack](https://github.com/CanRau/remix-new-age-stack) with Streaming 🥳

What worked for me was:

```javascript
[...document.querySelectorAll("html>*")]
  .filter((elem) => !elem.matches("head,body"))
  .forEach((s) => {
    s.parentNode?.removeChild(s);
  });
```

This removes anything added to html from extensions like grammarly. Also add `suppressHydrationWarning={true}` in the `html` and `body` tags in the case of grammarly. I faced this situation in production in an app created with `npx create-rsc-app@latest my-app --ssr` (it's a setup with RSC and SSR I've made myself).

**Comment by Mordred** [comment]
### Today I learned

Not every script tags and style tags comes from browser extensions. Kaspersky Internet Security -> Anti-Banner feature is injecting HTML tags on the network layer (with MiTM, but that is not relevant now :D). This effectively triggers hydration errors.

For now: I'm extending my `querySelectorAll` to

```js
  document
    .querySelectorAll(
      'html > *:not(body, head), script[src*="extension://"], link[href*="extension://"], script[src*="scr.kaspersky-labs.com"], link[src*="scr.kaspersky-labs.com"]',
    )
    .forEach((s) => {
      s.parentNode.removeChild(s);
    });

```

If this will not work I swear I will add `data-my-tag` to every `script`, `link`, and `style` tag which were generated by my server 😂 

There is also another problem. React is trying to recover but it replace all HTML elements with new ones. This breaks CSS-in-JS solutions e.g. Emotion, because they already have references to old elements where they are still injecting their styles. 

I'm using the MUI library in one of my projects and after the hydration error, all styles are gone.

There is an `onRecoverableError` but it is called after the DOM was changed by React. 

Can we have a callback that will be called before React tries to recover from the hydration mismatch? (@gnoff @gaearon)

I would like to make a snapshot of DOM for better debugging of what people have injected in their browsers. Also Emotion (or another) library can be reset (clear their elements refs) before react switch to client-side rendering (`An error occurred during hydration. The server HTML was replaced with client content in <#document>.`).

**Comment by giankotarola** [comment]
my hydration error got fixed with:
```
"react": "^18.3.0-canary-3d9b20132-20240124",
"react-dom": "^18.3.0-canary-3d9b20132-20240124",
```
For testing i added back the cloudflare email obfuscated script injection + enable back the chrome extensions that triggers the hydration issue for me, Apollo client dev tools / Requestly and have not seen the issue 🙌

**Comment by rcfrias** [comment]
I cant believe this is still a problem for the past 2 years.  I have LastPass installed, and even a Hello World app with ONE input tag type "text" and a placeholder "email" is causing the hydration issue. and no, suppressHydrationWarning does not fix the problem. (placed it everywhere without luck) :(

**Comment by HakanSungur** [comment]
I'm getting the same error and I'm looking for a solution.

**Comment by Omar-Abdul-Azeez** [comment]
injected elements with style attribute throw an ssr mismatch error... why is this not fixed yet...

**Comment by grundmanise** [comment]
With all due respect @Omar-Abdul-Azeez, but it's an OS project..

**Comment by johnlomat** [comment]
Can confirm LastPass browser extension causing the hydration mismatch.

**Comment by Coachonko** [comment]
> Can confirm LastPass browser extension causing the hydration mismatch.

Can you please tell me if this happens on 18.3.1 and on canary?

**Comment by yemi** [comment]
Seems react 19 (beta) _might_ fix this, for me at least. Renders nicely with browser addons without errors

**Comment by wilbertcaba** [comment]
> Can confirm LastPass browser extension causing the hydration mismatch.

Can confirm the same. Lastpass browser extension was causing error.

**Comment by digitaldjango** [comment]
Lastpass issue confirmed here as well. Took me a while to figure it out and ended up here..

**Comment by JamesLeBoeuf** [comment]
> Seems react 19 (beta) _might_ fix this, for me at least. Renders nicely with browser addons without errors

Tried this, unfortunately didn't work for me. Confirmed still an issue with LastPass browser extension causing it.

**Comment by allicanseenow** [comment]
Still seems to be an issue when testing with Remix and React 18.3.1. It happens when the browser tab has been inactive for some time and it is focused again.

**Comment by bialasky** [comment]
same happening with extension "I still don't care about cookies".

**Comment by triemstr** [comment]
I still get the warning with LastPass on the Oct 29 2024 React and React-Dom versions [19.0.0-rc-603e6108-20241029](https://www.npmjs.com/package/react-dom/v/19.0.0-rc-603e6108-20241029)...but I love the new messaging. It shows git-style "minus" signs to show the difference between the server and client renderings, not to mention way better explanations and formatting. I think it is worth showing a screenshot for those who don't have this level of detail in React 18. Kudos to developers to analyze the issue to this level of detail.

![image](https://github.com/user-attachments/assets/cab6fbc5-2096-4569-8905-7e8601c14dba)

**Comment by raphaelbadia** [comment]
if lastpass adds the tag suppressHydrationWarning I believe that the error could go away.

At least that's how [jam](https://jam.dev/s) works

![Screenshot 2024-11-04 at 13 42 18](https://github.com/user-attachments/assets/f0a2ca6b-29f1-47f6-90bb-255f1d1d49d6)

**Comment by karlhorky** [comment]
Hydration error seems to be happening also with [Video Speed Controller Chrome Extension](https://chromewebstore.google.com/detail/video-speed-controller/nffaoalbilbmmfgbnbgppjihopabppdk) with the Next.js version `next@15.0.4-canary.41` (React version `19.0.0-rc-de68d2f4-20241204`) in the reproduction template:

```
Hydration failed because the server rendered HTML didn't match the client. As a result this tree will be regenerated on the client. This can happen if a SSR-ed Client Component used
```

- https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template 

![Screenshot 2024-12-05 at 12 20 14](https://github.com/user-attachments/assets/bd61a1f5-5109-4b25-92ea-4752d7d8ea17)

![Screenshot 2024-12-05 at 12 20 18](https://github.com/user-attachments/assets/a42cfe93-15f1-47be-badb-214eba073318)

**Update:** New message here:

```
hook.js:608 A tree hydrated but some attributes of the server rendered HTML didn't match the client properties. This won't be patched up. This can happen if a SSR-ed Client Component used:

- A server/client branch `if (typeof window !== 'undefined')`.
- Variable input such as `Date.now()` or `Math.random()` which changes each time it's called.
- Date formatting in a user's locale which doesn't match the server.
- External changing data without sending a snapshot of it along with the HTML.
- Invalid HTML tag nesting.

It can also happen if the client has a browser extension installed which messes with the HTML before React loaded.

https://react.dev/link/hydration-mismatch

  ...
    <HotReload assetPrefix="" globalError={[...]}>
      <AppDevOverlay state={{nextId:1, ...}} dispatch={function bound dispatchReducerAction} globalError={[...]}>
        <AppDevOverlayErrorBoundary globalError={[...]} onError={function}>
          <ReplaySsrOnlyErrors>
          <DevRootHTTPAccessFallbackBoundary>
            <HTTPAccessFallbackBoundary notFound={<NotAllowedRootHTTPFallbackError>}>
              <HTTPAccessFallbackErrorBoundary pathname="/" notFound={<NotAllowedRootHTTPFallbackError>} ...>
                <RedirectBoundary>
                  <RedirectErrorBoundary router={{...}}>
                    <Head>
                    <RootLayout>
                      <html lang="en">
                        <body
-                         className="vsc-initialized"
                        >
                    ...
        ...
```


Also reported in the Video Speed Controller extension issues here:

- https://github.com/igrigorik/videospeed/issues/1255
- https://github.com/igrigorik/videospeed/issues/1160

**Comment by BSPREDDY** [comment]
<img width="960" alt="Screenshot 2024-12-21 145936" src="https://github.com/user-attachments/assets/2a8162b7-9eb0-4e10-bfcb-b45c1bdffcfd" />
how to solve this problem

**Comment by rapidstackjobs** [comment]
My Bookmark Sidebar chrome extension was the culprit in case someone else has that extension installed. I disabled it during development.

**Comment by thecodermehedi** [comment]
In my case [Grammerly](https://addons.mozilla.org/en-US/firefox/addon/grammarly-1/) was the culprit.

**Comment by javadazizmohamad** [comment]
یادگیری وآموزش

**Comment by madfcat** [comment]
Also, happens to next 14.2.15., next-international 1.3.1, Google Chrome "Always translate" option in dev and prod.

**Comment by alvyynm** [comment]
Happened to me today. Dark Reader was the culprit.

**Comment by javadazizmohamad** [comment]
من،بروزرسانی،مخزنم،به،پایان،رسیده،بود،بخاطر،همین،ازشما،طلب،کمک،کردم

**Comment by karlhorky** [comment]
It looks like Next.js 16.0.0+ no longer reports hydration errors on `html` element 😳

- https://github.com/igrigorik/videospeed/issues/1417#issuecomment-3896586584

<img width="1435" height="611" alt="Image" src="https://github.com/user-attachments/assets/bcec909b-87f3-4c50-8980-191acda51694" />

Next.js 15.5.12 to contrast:

<img width="1436" height="856" alt="Image" src="https://github.com/user-attachments/assets/7c5c99de-0303-4428-a03a-df5ddd5191ca" />

Seems like this was a change in Next.js v16 or React...?

**Comment by karlhorky** [comment]
@eps1lon mentioned offline that this was a change in React:

> We changed in React 19 how hydration works for `html`, `head` and `body`.
> 
> We made those a lot more resistant to handle 3rd party code

So maybe this issue should be closed now?

@yongdamsh or anyone else in the thread: are you still experiencing problems with some browser extensions?

**Comment by grundmanise** [comment]
@karlhorky I think that was mentioned here: https://react.dev/blog/2024/12/05/react-19#compatibility-with-third-party-scripts-and-extensions

**Comment by karlhorky** [comment]
@grundmanise I have seen this, yes, and at first it appears to be related.

But as you can see in [my post above](https://github.com/facebook/react/issues/24430#issuecomment-3896653927), it remained a problem in Next.js long after the React 19 release (latest Next.js v15 release still shows an error with the Video Speed Controller browser extension).

So it wasn't fixed in React 19 - it was fixed much later (although I haven't yet found the exact PRs where the changes were made).

**Comment by SukkaW** [comment]
<img width="1408" height="1110" alt="Image" src="https://github.com/user-attachments/assets/f5d0a79d-9093-4a19-9089-2603c510afcd" />

Here is the situation of the "Grammarly" browser extension (which I am using), they *fxxking* added their attributes to `<body>`.

**Comment by karlhorky** [comment]
@SukkaW Ah, thanks for the note. Too bad!

I guess this issue should still be kept open then, to track the browser extensions which cause React hydration mismatch problems.

**Comment by BSPREDDY** [comment]
Look at this part:

data-new-gr-c-s-check-loaded="14.1275.0"
data-gr-ext-installed=""

These attributes are *NOT from your code*.

They are added by the *Grammarly browser extension*.

So the flow becomes:


   1.

   Server sends HTML
   2.

   Grammarly extension *modifies the HTML*
   3.

   React tries to hydrate
   4.

   HTML changed → *hydration mismatch error*

Solution 1 (Recommended)

Disable *Grammarly extension* for localhost.

Steps:

   1.

   Open Chrome Extensions
   2.

   Find *Grammarly*
   3.

   Disable it for your dev site

OR open site in *Incognito mode*.
------------------------------
Solution 2

Run your app in *Incognito* where extensions are disabled.
------------------------------
Solution 3 (If issue from code)

Hydration errors also happen if you use things like:

Math.random()
Date.now()
typeof window !== "undefined"

inside server-rendered components.

But *your case is clearly a browser extension*.
